// @bit/ofbtech.react-boilerplate.template/dist/template/app/containers/App/action.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.updatePageInfo = updatePageInfo;
exports.notifySuccess = notifySuccess;
exports.notifyError = notifyError;
exports.notifyWarn = notifyWarn;

var _reactToastify = require("react-toastify");

var _constants = require("./constants");

/*
 * App Actions
 *
 */
function updatePageInfo(pageInfo) {
  return {
    type: _constants.PAGE_INFO,
    pageInfo: pageInfo
  };
}

function notifySuccess() {
  var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'Success';

  _reactToastify.toast.success(msg);

  return {
    type: _constants.NOTIFICATION_SUCCESS
  };
}

function notifyError() {
  var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {
    message: 'Error'
  };

  _reactToastify.toast.error(msg.message);

  return {
    type: _constants.NOTIFICATION_ERROR
  };
}

function notifyWarn() {
  var msg = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : 'Warning';

  _reactToastify.toast.warn(msg);

  return {
    type: _constants.NOTIFICATION_WARN
  };
}
