// @bit/ofbtech.react-boilerplate.template/dist/template/app/components/form/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _tcombForm = _interopRequireDefault(require("tcomb-form"));

var _CheckBox = _interopRequireDefault(require("./CheckBox"));

var _TextBox = _interopRequireDefault(require("./TextBox"));

var _Select = _interopRequireDefault(require("./Select"));

var _Radio = _interopRequireDefault(require("./Radio"));

var _Date = _interopRequireDefault(require("./Date"));

var _Scale = _interopRequireDefault(require("./Scale"));

var _common = _interopRequireDefault(require("../../utils/common"));

var _regex = _interopRequireDefault(require("../../utils/regex"));

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

var _default = {
  templates: {
    checkbox: _CheckBox["default"],
    radio: _Radio["default"],
    select: _Select["default"],
    textbox: _TextBox["default"],
    date: _Date["default"],
    scale: _Scale["default"]
  },
  transformers: {
    listTransformer: {
      format: function format(value) {
        return Array.isArray(value) ? value.join(',') : value;
      },
      parse: function parse(str) {
        return _common["default"].isStringAndNotEmpty(str) ? str.split(',') : null;
      }
    },
    dateTransformer: {
      format: function format(value) {
        if (!_common["default"].isUndefinedOrNullOrEmpty(value)) {
          return new Date(value);
        }

        return value;
      },
      parse: function parse(value) {
        if (!_common["default"].isUndefinedOrNullOrEmpty(value)) {
          return new Date(value).getTime();
        }

        return null;
      }
    }
  },
  refinements: {
    emailEmptyAllowed: _tcombForm["default"].refinement(_tcombForm["default"].maybe(_tcombForm["default"].Str), function (s) {
      return _common["default"].isUndefinedOrNullOrEmpty(s) || _regex["default"].plainEmail.test(s);
    }),
    email: _tcombForm["default"].subtype(_tcombForm["default"].Str, function (s) {
      return _regex["default"].email.test(s);
    }),
    mobile: _tcombForm["default"].subtype(_tcombForm["default"].Str, function (s) {
      return _regex["default"].mobile.test(s);
    }),
    pinCode: _tcombForm["default"].refinement(_tcombForm["default"].Str, function (s) {
      return _regex["default"].pincode.test(s);
    }),
    year: _tcombForm["default"].refinement(_tcombForm["default"].Str, function (s) {
      return _regex["default"].year.test(s);
    }),
    pancard: _tcombForm["default"].refinement(_tcombForm["default"].Str, function (s) {
      return _regex["default"].pancard.test(s);
    }),
    gst: _tcombForm["default"].refinement(_tcombForm["default"].maybe(_tcombForm["default"].Str), function (s) {
      return !_common["default"].isUndefinedOrNullOrEmpty(s) ? _regex["default"].gst.test(s) : true;
    }),
    organisationName: _tcombForm["default"].refinement(_tcombForm["default"].Str, function (s) {
      return _regex["default"].organisationName.test(s);
    }),
    accountName: _tcombForm["default"].refinement(_tcombForm["default"].Str, function (s) {
      return _regex["default"].accountName.test(s);
    }),
    hasPositiveNumericValue: _tcombForm["default"].refinement(_tcombForm["default"].Number, function (s) {
      return s > 0;
    }),
    percentage: _tcombForm["default"].refinement(_tcombForm["default"].Str, function (s) {
      return !_common["default"].isUndefinedOrNullOrEmpty(s) && _regex["default"].percentage.test(s);
    }),
    turnover: _tcombForm["default"].refinement(_tcombForm["default"].Str, function (s) {
      return _regex["default"].turnover.test(s);
    }),
    validOtp: _tcombForm["default"].refinement(_tcombForm["default"].Str, function (s) {
      return _regex["default"].otp.test(s);
    })
  },
  helper: {
    build: function build(options, isFormSubmit) {
      var newOption = options;
      Object.keys(newOption.fields).forEach(function (field) {
        if (_common["default"].isUndefinedOrNullOrEmpty(newOption.fields[field].attrs)) {
          newOption.fields[field].attrs = {};
        }

        newOption.fields[field].attrs.isFormSubmit = isFormSubmit;
      });
      return newOption;
    }
  }
};
exports["default"] = _default;
