// @bit/ofbtech.react-boilerplate.template/dist/template/app/containers/ToDo/constants.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.TODO_FILTERS = exports.FILTER_COMPLETED = exports.FILTER_ACTIVE = exports.FILTER_ALL = exports.ADD_TODO = exports.SET_SEARCH_QUERY = exports.UPDATE_TODO = exports.CHANGE_FILTER = exports.CHANGE_STATUS = void 0;

var _TODO_FILTERS;

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

/*
 *
 * ToDo constants
 *
 */
var CHANGE_STATUS = 'app/ToDo/CHANGE_STATUS';
exports.CHANGE_STATUS = CHANGE_STATUS;
var CHANGE_FILTER = 'app/ToDo/CHANGE_FILTER';
exports.CHANGE_FILTER = CHANGE_FILTER;
var UPDATE_TODO = 'app/ToDo/UPDATE_STATUS';
exports.UPDATE_TODO = UPDATE_TODO;
var SET_SEARCH_QUERY = 'app/ToDo/SET_SEARCH_QUERY';
exports.SET_SEARCH_QUERY = SET_SEARCH_QUERY;
var ADD_TODO = 'app/ToDo/ADD_TODO';
exports.ADD_TODO = ADD_TODO;
var FILTER_ALL = 'all';
exports.FILTER_ALL = FILTER_ALL;
var FILTER_ACTIVE = 'active';
exports.FILTER_ACTIVE = FILTER_ACTIVE;
var FILTER_COMPLETED = 'completed';
exports.FILTER_COMPLETED = FILTER_COMPLETED;
var TODO_FILTERS = (_TODO_FILTERS = {}, _defineProperty(_TODO_FILTERS, FILTER_ALL, 'All'), _defineProperty(_TODO_FILTERS, FILTER_ACTIVE, 'Active'), _defineProperty(_TODO_FILTERS, FILTER_COMPLETED, 'Completed'), _TODO_FILTERS);
exports.TODO_FILTERS = TODO_FILTERS;
