// @bit/ofbtech.react-boilerplate.template/dist/template/app/containers/ToDo/components/TodoFilter/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = ToDoFilter;

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _constants = require("../../constants");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

/**
 *
 * ToDo - Filter
 *
 */
function ToDoFilter(_ref) {
  var filter = _ref.filter,
      changeFilter = _ref.changeFilter;

  var getClass = function getClass(key) {
    return key === filter ? 'selected' : '';
  };

  return _react["default"].createElement("ul", {
    className: "filters list-unstyled clearfix"
  }, Object.keys(_constants.TODO_FILTERS).map(function (key) {
    return _react["default"].createElement("li", {
      key: key
    }, _react["default"].createElement("button", {
      type: "button",
      onClick: function onClick() {
        return changeFilter(key);
      },
      className: getClass(key)
    }, _constants.TODO_FILTERS[key]));
  }));
}

ToDoFilter.propTypes = {
  changeFilter: _propTypes["default"].func,
  filter: _propTypes["default"].string
};
ToDoFilter.defaultProps = {
  changeFilter: function changeFilter() {},
  filter: null
};
