// @bit/ofbtech.react-boilerplate.template/dist/template/internals/webpack/webpack.dev.babel.js
"use strict";

/**
 * DEVELOPMENT WEBPACK CONFIGURATION
 */
var path = require('path');

var webpack = require('webpack');

var HtmlWebpackPlugin = require('html-webpack-plugin');

var CircularDependencyPlugin = require('circular-dependency-plugin');

var TimeFixPlugin = require('time-fix-plugin');

module.exports = require('./webpack.base.babel')({
  mode: 'development',
  // Add hot reloading in development
  entry: [require.resolve('react-app-polyfill/ie11'), 'webpack-hot-middleware/client?reload=true', path.join(process.cwd(), 'app/app.js') // Start with js/app.js
  ],
  // Don't use hashes in dev mode for better performance
  output: {
    filename: '[name].js',
    chunkFilename: '[name].chunk.js'
  },
  optimization: {
    splitChunks: {
      chunks: 'all'
    }
  },
  // Add development plugins
  plugins: [new TimeFixPlugin(), new webpack.HotModuleReplacementPlugin(), // Tell webpack we want hot reloading
  new HtmlWebpackPlugin({
    inject: true,
    // Inject all files that are generated by webpack, e.g. bundle.js
    template: 'app/index.html'
  }), new CircularDependencyPlugin({
    exclude: /a\.js|node_modules/,
    // exclude node_modules
    failOnError: false // show a warning when there is a circular dependency

  })],
  // Emit a source map for easier debugging
  // See https://webpack.js.org/configuration/devtool/#devtool
  devtool: 'eval-source-map',
  performance: {
    hints: false
  }
});
